pipeline-model-definition/src/main/groovy/org/jenkinsci/plugins/pipeline/modeldefinition/Utils.groovy

```diff
@@ -205,6 +212,71 @@ public class Utils {
     }
 
     /**
+     * Returns true if we're currently nested under a stage.
+     *
+     * @return true if we're in a stage and false otherwise
+     */
+    static boolean withinAStage() {
+        CpsThread thread = CpsThread.current()
+        CpsFlowExecution execution = thread.execution
+
+        LinearBlockHoppingScanner scanner = new LinearBlockHoppingScanner();
+
+        FlowNode stageNode = scanner.findFirstMatch(execution, isStageWithOptionalName())
+
+        return stageNode != null
+    }
+
+    private static Predicate<FlowNode> isStageWithOptionalName(final String stageName = null) {
+        return new Predicate<FlowNode>() {
+            @Override
+            boolean apply(@Nullable FlowNode input) {
+                return input != null &&
+                    input instanceof StepStartNode &&
+                    ((StepStartNode)input).descriptor instanceof StageStep.DescriptorImpl &&
+                    (stageName == null || input.displayName?.equals(stageName))
+            }
+        }
+    }
+
+    private static void markStageWithTag(String stageName, String tagName, String tagValue) {
+        CpsThread thread = CpsThread.current()
+        CpsFlowExecution execution = thread.execution
+
+        LinearBlockHoppingScanner scanner = new LinearBlockHoppingScanner();
+
+        FlowNode currentNode = scanner.findFirstMatch(execution, isStageWithOptionalName(stageName))
+
+        TagsAction tagsAction = currentNode.getAction(TagsAction.class)
+        if (tagsAction == null) {
+            tagsAction = new TagsAction()
+            tagsAction.addTag(tagName, tagValue)
+            currentNode.addAction(tagsAction)
```

Comment:
@rsandell FYI if curious, b/c `addAction` in `FlowNode` does it automagically by override of `getActions()` with a custom list whose mutators trigger `save()`